File: Golang/Replace Words.go
func replaceWords(dictionary []string, sentence string) string {
    roots := make(map[string]bool)
    for _, root := range dictionary {
        roots[root] = true
    }

    words := strings.Split(sentence, " ")
    var result []string

    for _, word := range words {
        found := false
        for i := 1; i <= len(word); i++ {
            prefix := word[:i]
            if _, exists := roots[prefix]; exists {
                result = append(result, prefix)
                found = true
                break
            }
        }

        if !found {
            result = append(result, word)
        }
    }

    return strings.Join(result, " ")
}

File: Python3/Replace Words.py
#1
class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        rootset = set(dictionary)

        def replace(word):
            for i in range(1, len(word)):
                if word[:i] in rootset:
                    return word[:i]
            return word

        return " ".join(map(replace, sentence.split()))

#2
class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        trie = {}
        
        for word in dictionary:
            t = trie
            for ch in word:
                if ch not in t:
                    t[ch] = {}
                t = t[ch]
            t['#'] = '#'

        def process(string):
            t = trie
            for i,ch in enumerate(string):
                if ch not in t:
                    break
                t = t[ch]
                if '#' in t:
                    return string[:i+1]
            return string
        
        return ' '.join(map(process, sentence.split()))